lib/id3c/cli/command/redcap_sync.py

```diff
@@ -0,0 +1,327 @@
+
+    deletion_count = 0
+    for encounter in del_encounters:
+        encounter_id = encounter.encounter_id
+        LOG.info(
+            f"Processing deleted encounter {encounter_id} from project {project.id}"
+        )
+
+        with db.savepoint(f"encounter {encounter.encounter_id}"):
+            delete_linked_encounter_records(db, encounter_id, encounter.individual_id)
+            deletion_count += 1
+
+        LOG.info(
+            f"Finished processing deleted encounter {encounter_id} from project {project.id}"
+        )
+
+    LOG.info(
+        f"Synced {len(deletion_events)} REDCap deletion events with ID3C. {deletion_count} ID3C encounters and associated data were removed."
+    )
+
+
+def delete_linked_encounter_records(
+    db: DatabaseSession, encounter_id: int, individual_id: int
+):
+    """
+    Deletes data associated with the linked `encounter_id` from the provided `db`.
+    If other encounter_ids are linked to the same record as the provided `encounter_id`,
+    don't delete those records.
+
+    This function may alter data from the following tables and should be used with care:
+    - `warehouse.encounter`
+    - `warehouse.encounter_location`
+    - `warehouse.individual`
+    - `warehouse.location`
+    - `warehouse.sample`
+    - `warehouse.presence_absence`
+    """
+    LOG.info(f"Deleting all relational encounter data for encounter {encounter_id}")
+
+    # Take a snapshot of relational location data that is safe to delete before we delete our encounter.
+    location_ids_unique_to_this_encounter = db.fetch_all(
+        f"""
+          SELECT
+            location_id
+          FROM
+            warehouse.encounter_location t1
+          WHERE
+            encounter_id = {encounter_id}
+            AND location_id NOT IN
+              (
+                SELECT
+                  location_id
+                FROM
+                  warehouse.encounter_location
+                WHERE
+                  location_id = t1.location_id
+                  AND encounter_id <> {encounter_id}
+              )
+        """
+    )
+
+    deleted_encounter_loc_rows = delete_encounter_location(db, encounter_id)
+    LOG.debug(
+        f"Deleted {deleted_encounter_loc_rows} rows from `encounter_location` associated "
+        + f"with `encounter_id` = {encounter_id}"
+    )
+
+    sample_ids_associated_with_this_encounter = db.fetch_all(
+        f"SELECT encounter_id, sample_id FROM warehouse.sample WHERE encounter_id = {encounter_id}"
+    )
+    for sample in sample_ids_associated_with_this_encounter:
+        deleted_pa_rows = delete_presence_absence_by_sample(db, sample.sample_id)
+        deleted_sample_rows = delete_sample(db, sample.sample_id)
+        LOG.debug(
+            f"Deleted {deleted_pa_rows} rows from `presence_absence` and {deleted_sample_rows} from `sample` associated "
+            + f"with `encounter_id` = {sample.encounter_id} and `sample_id` = {sample.sample_id}"
+        )
```

Comment:
We probably want to avoid deleting sample records if they were also received from LIMS, but it might make sense to scrub the record and re-sync from LIMS to make sure we don't have anything from REDCap included.

Reply:
Yeah good thought. Going to follow up with this on Slack

Reply:
Handling this by deleting records with only `note` and `coding` keys in their details. This indicates they have a provenance of REDCap and were not included on any LIMS / AQ sheet incident reports.